File: java-automation/src/test/java/com/myecommerce/automation/driver/api/OkHttpHttpPort.java
package com.myecommerce.automation.driver.api;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.myecommerce.automation.driver.ports.HttpPort;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

import java.io.IOException;
import java.util.List;

public final class OkHttpHttpPort implements HttpPort {

    private static final ObjectMapper MAPPER = new ObjectMapper()
        .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
        .setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE);

    private final OkHttpClient client = new OkHttpClient();
    private final String baseUrl;

    public OkHttpHttpPort(String baseUrl) {
        this.baseUrl = baseUrl;
    }

    @Override
    public <T> T getAs(String path, Class<T> type) {
        String json = get(path, null, null);
        try {
            return MAPPER.readValue(json, type);
        } catch (Exception e) {
            throw new RuntimeException("Failed to deserialize " + path, e);
        }
    }

    @Override
    public <T> List<T> getListAs(String path, Class<T> elementType) {
        return deserializeList(get(path, null, null), elementType);
    }

    @Override
    public <T> List<T> getListWithQueryAs(String path, String paramName, String paramValue, Class<T> elementType) {
        return deserializeList(get(path, paramName, paramValue), elementType);
    }

    private String get(String path, String paramName, String paramValue) {
        HttpUrl.Builder urlBuilder = HttpUrl.parse(baseUrl + path).newBuilder();
        if (paramName != null) urlBuilder.addQueryParameter(paramName, paramValue);
        Request request = new Request.Builder().url(urlBuilder.build()).build();
        try (Response response = client.newCall(request).execute()) {
            if (response.code() != 200) {
                throw new RuntimeException("Expected 200 but got " + response.code() + " for " + path);
            }
            return response.body().string();
        } catch (IOException e) {
            throw new RuntimeException("HTTP request failed: " + path, e);
        }
    }

    private <T> List<T> deserializeList(String json, Class<T> elementType) {
        var listType = MAPPER.getTypeFactory().constructCollectionType(List.class, elementType);
        try {
            return MAPPER.readValue(json, listType);
        } catch (Exception e) {
            throw new RuntimeException("Failed to deserialize list", e);
        }
    }
}

File: kotlin-automation/src/test/kotlin/com/myecommerce/automation/driver/api/OkHttpHttpPort.kt
package com.myecommerce.automation.driver.api

import com.fasterxml.jackson.databind.DeserializationFeature
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.PropertyNamingStrategies
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import com.myecommerce.automation.driver.ports.HttpPort
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request

class OkHttpHttpPort(private val baseUrl: String) : HttpPort {

    private val client = OkHttpClient()
    private val mapper = ObjectMapper()
        .registerKotlinModule()
        .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
        .setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE)

    override fun <T : Any> getAs(path: String, type: Class<T>): T =
        mapper.readValue(get(path), type)

    override fun <T : Any> getListAs(path: String, elementType: Class<T>): List<T> =
        deserializeList(get(path), elementType)

    override fun <T : Any> getListWithQueryAs(path: String, paramName: String, paramValue: String, elementType: Class<T>): List<T> =
        deserializeList(get(path, paramName to paramValue), elementType)

    private fun get(path: String, query: Pair<String, String>? = null): String {
        val url = "$baseUrl$path".toHttpUrl().newBuilder()
            .apply { query?.let { addQueryParameter(it.first, it.second) } }
            .build()
        val response = client.newCall(Request.Builder().url(url).build()).execute()
        check(response.code == 200) { "Expected 200 but got ${response.code} for $path" }
        return response.body?.string() ?: ""
    }

    private fun <T : Any> deserializeList(json: String, elementType: Class<T>): List<T> {
        val listType = mapper.typeFactory.constructCollectionType(List::class.java, elementType)
        return mapper.readValue(json, listType)
    }
}
